### PR TITLE
feat(dev): ui hot reloading in development

### PR DESCRIPTION
**Pull request by simonheys**
Adds very simple hot reloading to the UI. When used, changing any file in the 'ui' folder will cause the extension UI to reload and reinstate the previous browser route. This means you can see changes without having to close and re-open the extension.

Works by adding a simple local WebSocket server which exchanges messages with the in-page script, which can then send messages in turn to the UI.

https://user-images.githubusercontent.com/175607/179465624-9700d319-e20c-47e5-9bc1-92933ab7c356.mov

### CODE REVIEW

File: packages/extension/package.json [hunk 1]
```diff
@@ -16,9 +16,12 @@
     "@types/react-measure": "^2.0.8",
     "@types/styled-components": "^5.1.25",
     "@types/url-join": "^4.0.1",
+    "@types/ws": "^8.5.3",
     "@typescript-eslint/eslint-plugin": "^5.10.1",
     "@typescript-eslint/parser": "^5.10.1",
     "@vitejs/plugin-react": "^2.0.0",
+    "chokidar": "^3.5.2",
```
Comment (simonheys) [inline packages/extension/package.json:23]: For watching local file systems in local development

File: packages/extension/package.json [hunk 2]
```diff
@@ -16,9 +16,12 @@
...
+    "concurrently": "^7.2.2",
```
Comment (simonheys) [inline packages/extension/package.json:24]: For running multiple concurrent commands in a single script

File: packages/extension/package.json [hunk 3]
```diff
@@ -43,14 +46,16 @@
     "vitest": "^0.18.0",
     "wait-for-expect": "^3.0.2",
     "webpack": "^5.62.1",
-    "webpack-cli": "^4.9.1"
+    "webpack-cli": "^4.9.1",
+    "ws": "^8.8.1"
```
Comment (simonheys) [inline packages/extension/package.json:50]: WebSocket server for local node

File: packages/extension/package.json [hunk 4]
```diff
@@ -43,14 +46,16 @@
...
   },
   "scripts": {
     "build": "NODE_ENV=production webpack",
     "build:sourcemaps": "NODE_ENV=production GEN_SOURCE_MAPS=true webpack",
     "start": "webpack",
-    "dev": "webpack --color --watch",
-    "dev:ui": "SHOW_DEV_UI=true webpack --color --watch",
+    "dev": "concurrently -k -r \"webpack --color --watch\" \"yarn dev:hot-reload-server\"",
```
Comment (simonheys) [inline packages/extension/package.json:56]: Runs the hot reload server in parallel with existing webpack

File: packages/extension/src/inpage/index.ts [hunk 5]
```diff
@@ -72,3 +73,7 @@ window.addEventListener(
     }
   },
 )
+
+if (IS_DEV) {
+  require("./hotReload")
```
Comment (simonheys) [inline packages/extension/src/inpage/index.ts:78]: Ensures no dev implementation leaks into production code

File: packages/extension/src/shared/utils/dev.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export const IS_DEV = process.env.NODE_ENV === "development"
+
+export const HOT_RELOAD_PORT = 9001
```
Comment (simonheys) [inline packages/extension/src/shared/utils/dev.ts:3]: Arbitrary number, but normal webpack hot reload would be on port 9000... so this is +1

**Comment by dhruvkelawala** [comment]
Conflicts

**Review by janek26 — APPROVED** [review]
conflicts

**Pull request merged**